### DISCUSSION TITLE
validate_on_init or __pydantic_run_validation__ doesn't work

### DISCUSSION THREAD

**Discussion by PoojaAg18** [pydantic/pydantic#4503] [Question] [answered]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

On trying to run below code, it results into validation errors because i am passing string type value for port rather then list -

```
pydantic.error_wrappers.ValidationError: 1 validation error for MySQLConfig
port
value is not a valid list (type=type_error.list)
```

I am trying to use validate_on_init or pydantic_run_validation to avoid this pydantic validation errors. but this doesn't work

Can you help me finding a way to avoid pydantic init validation errors when we pass different field type value? this works with stdlib dataclasses?

### Example Code

```Python
@dataclass()
class MySQLConfig:
    driver: Dict[str, str] = field(default_factory=dict)
    port: List[str] = field(default_factory=list) 
    password: int = 0
    
    @validator("password")
    def validate_supported_password(password: int, values, **kwargs):
      if password == 100:
          raise ValueError(f"'password' cannot be 100 got: {password}")
      return password

config = MySQLConfig(driver={"test":"omry"}, port="13", password=3, validate_on_init=False) 
config = MySQLConfig(driver={"test":"omry"}, port="13", password=3, __pydantic_run_validation__=False)
config = MySQLConfig(driver={"test":"omry"}, port="13", password=3, validate_on_init=False).__pydantic_run_validation__=False
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.2
pydantic compiled: True
install path: /home/runner/VainSuperficialEmulators/venv/lib/python3.8/site-packages/pydantic
               python version: 3.8.12 (default, Aug 30 2021, 16:42:10)  [GCC 10.3.0]
                     platform: Linux-5.15.0-1017-gcp-x86_64-with
optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [X] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [X] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [X] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [X] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Settings Management](https://pydantic-docs.helpmanual.io/usage/settings/)
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by PoojaAg18** [comment]
@PrettyWood  @samuelcolvin  I have submitted a new issue with better description to tell about my issue. as I was not able to reopen last issue [https://github.com/pydantic/pydantic/issues/4496] . Please guide me if I am doing something wrong in the code

**Comment by samuelcolvin** [comment]
This is a question, not a bug. I'll try to reply and help tomorrow.

The instructions on questions vs. bugs are clear afaik, please try to follow them.

**Comment by samuelcolvin** [comment]
Honestly, I don't understand what you're trying to do.

But you definitely can pass config arguments like `validate_on_init` when initializing dataclasses (or models).

**Reply by PoojaAg18** [reply]
Thanks for your reply @samuelcolvin 

I am trying to initialize dataclass by passing field values with different types.

**For example** - in above case, I have tried to pass "str" value for port rather then "List[str]". This throws pydantic validation error but when we use stdlib dataclass it simply converts the field from List[str] to str without throwing any error.

I was trying to get similar behavior, and use **validate_on_init** but it didn't help avoiding pydantic validation errors here. Is there a way to avoid the below errors -

```
pydantic.error_wrappers.ValidationError: 1 validation error for MySQLConfig
port
value is not a valid list (type=type_error.list)
```

**Comment by samuelcolvin** [comment, marked as answer]
I don't know where you got `validate_on_init` there's no such argument for pydantic.

If you don't want validation on `__init__`, don't use pydantic.

> when we use stdlib dataclass it simply converts the field from List[str] to str without throwing any error.

No. There's no "converting fields" in python dataclasses, just that type are not validated at run time when using standard dataclasses.

If you want validation, use pydantic. If you don't want validation use standard dataclasses.
